ipykernel/inprocess/client.py

```diff
@@ -173,8 +175,8 @@ def _dispatch_to_kernel(self, msg):
         stream = kernel.shell_stream
         self.session.send(stream, msg)
         msg_parts = stream.recv_multipart()
-        kernel.dispatch_shell(msg_parts)
-
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(kernel.dispatch_shell(msg_parts))
```

Comment:
- `dispatch_shell` is now an async function.
 - before, it was a tornado coroutine, that was called synchronously in this function.

Reply:
cc @minrk. Let me know what you think about this.

(I think the previous behavior was not correct)

Reply:
This assumes asyncio isn't running, which I guess is true?

To preserve the tornado behavior would be to use `asyncio.ensure_future(kernel.dispatch_shell(msg_parts))` which assumes the loop *is* (or at least will be) running.

Maybe add a check for if the current loop is running, and call `run_until_complete` if it's not, and `ensure_future` if it is.

The main difference between tornado and asyncio coroutines is that calling a tornado coroutine actually begins the call, whereas asyncio only creates a coroutine without starting any execution at all. `ensure_future` puts the coroutine on the loop.

Reply:
> This assumes asyncio isn't running, which I guess is true?

Absolutely. The loop is not running.

> The main difference between tornado and asyncio coroutines is that calling a tornado coroutine actually begins the call,

Yes - and the inprocess kernel is taking advantage of it, but never starting the loop! I think it probably a buggy behavior for the inprocess kernel.

Reply:
This is the last outstanding review comment, but I am not quite sure how to act upon it. The event loop is not running in the in-process kernel, however, we could be in a context of a larger application that has an event loop and we can't nest them...

Is the use case you are trying to solve?

In any case, unless we can nest event loop, the behavior will not be exactly the same... Hence I am not sure we should do anything about it.

Reply:
Adding some notes from testing - master definitely doesn't work if the eventloop is running, but it's less clear because tornado's generator-based coroutines execute up to the first yield (which is always the whole thing if the loop isn't running), whereas asyncio never runs anything when the loop isn't running. The code before accidentally assumes the loop isn't runnig, but the change here makes that explicit. It's not a new regression, but it make clearer that something that's never worked won't work. Something we can fix if people need it, but not in scope for this PR.